src/bin/mina-indexer-server.rs

```diff
@@ -25,79 +25,166 @@ struct ServerArgs {
+pub struct IndexerConfiguration {
+    root_hash: BlockHash,
+    startup_dir: PathBuf,
+    watch_dir: PathBuf,
+    store_dir: PathBuf,
+    log_file: PathBuf,
+    genesis_ledger: GenesisRoot,
+    log_stdout: bool,
+}
+
+#[instrument]
+pub async fn parse_command_line_arguments() -> anyhow::Result<IndexerConfiguration> {
+    event!(Level::INFO, "parsing ServerArgs");
     let args = ServerArgs::parse();
-    let genesis_ledger = match ledger::genesis::parse_file(&args.genesis_ledger).await {
-        Ok(genesis_root) => Some(genesis_root.ledger),
-        Err(e) => {
-            eprintln!(
-                "Unable to parse genesis ledger at {}: {}! Exiting.",
-                args.genesis_ledger.display(),
-                e
+    let root_hash = BlockHash(args.root_hash);
+    let startup_dir = args.startup_dir;
+    let watch_dir = args.watch_dir;
+    let store_dir = args.store_dir;
+    let log_dir = args.log_dir;
+    let log_stdout = args.log_stdout;
+    event!(Level::INFO, "parsing GenesisLedger file");
+    match ledger::genesis::parse_file(&args.genesis_ledger).await {
+        Err(err) => {
+            event!(
+                Level::ERROR,
+                reason = "unable to parse GenesisLedger",
+                error = err.to_string(),
+                path = &args.genesis_ledger.display().to_string()
             );
-            process::exit(100)
+            Err(err)
         }
-    };
+        Ok(genesis_ledger) => {
+            event!(
+                Level::INFO,
+                "GenesisLedger parsed {}",
+                args.genesis_ledger.display().to_string()
+            );
 
-    let root_hash = BlockHash(args.root_hash);
-    let logs_dir = args.logs_dir;
-    let store_dir = args.store_dir;
-    let mut indexer_state = mina_indexer::state::IndexerState::new(
+            let mut log_number = 0;
+            let mut log_file = format!("{}/mina-indexer-log-{}", log_dir.display(), log_number);
+            while tokio::fs::metadata(&log_file).await.is_ok() {
+                log_number += 1;
+                log_file = format!("{}/mina-indexer-log-{}", log_dir.display(), log_number);
+            }
+            let log_file = PathBuf::from(&log_file);
+
+            Ok(IndexerConfiguration {
+                root_hash,
+                startup_dir,
+                watch_dir,
+                store_dir,
+                log_file,
+                log_stdout,
+                genesis_ledger,
+            })
+        }
+    }
+}
+
+#[instrument]
+#[tokio::main]
+async fn main() -> Result<(), anyhow::Error> {
+    event!(Level::INFO, "started mina-indexer-server");
+    let IndexerConfiguration {
         root_hash,
+        startup_dir,
+        watch_dir,
+        store_dir,
+        log_file,
+        log_stdout,
         genesis_ledger,
-        store_dir.clone().as_deref(),
-    )?;
+    } = parse_command_line_arguments().await?;
+
+    // TODO: Improve Logging Subscribers
```

Comment:
If it's a TODO create an Asana task for it or else it will be lost

Reply:
https://app.asana.com/0/1203669416123378/1204789093923684/f